src/c2goto/library/builtin_libs.c

```diff
@@ -6,3 +6,268 @@ int __ESBMC_sync_fetch_and_add(int *ptr, int value)
   __ESBMC_atomic_end();
   return initial;
 }
+
+#if __ESBMC_CHERI__ == 128
+
+// #include <stdint.h>
+// #include <stddef.h>
+// #include <cheri.h>
+
+#if 1
+
+#ifdef __ESBMC_CHERI_MORELLO__
+#define CC_IS_MORELLO
+#endif
+
+#if !defined(cheri_debug_assert)
+/* Disable cheri-compressed-cap's debug assertions since they assert that
+ * base <= top in compute_base_top, which the comment above it admits is not
+ * always true. */
+#if 1
+#ifndef assert
+#define assert(expr) __ESBMC_assert(expr, "builtin libc assertion")
+#endif
+#define cheri_debug_assert(...)
+#else
+#define cheri_debug_assert(...)                                                \
+  __ESBMC_assert(__VA_ARGS__, "cheri-compressed-cap internal assertion")
+#endif
+#endif
+#include <cheri_compressed_cap.h>
+
+#ifdef __ESBMC_CHERI_MORELLO__
+#undef CC_IS_MORELLO
+#endif
+
+#include <cheri/cheric.h>
+// #include <assert.h>
+
+uint64_t __esbmc_clzll(uint64_t v)
+{
+  for(int i = 0; i < 64; i++, v <<= 1)
+    if(v & (uint64_t)1 << 63)
+      return i;
+  __ESBMC_assert(0, "clzll is not defined for zero arguments");
+  return nondet_uint64();
+}
+
+union __esbmc_cheri_cap128
+{
+  void *__capability cap;
+  // struct { __UINT64_TYPE__ pesbt, cursor; };
+  struct
+  {
+    __UINT64_TYPE__ cursor, pesbt;
+  };
+};
+
+__SIZE_TYPE__ __esbmc_cheri_base_get(void *__capability cap)
```

Comment:
Shall we add a short documentation about this method? You could use the doxygen style here.